File: packages/goblins/testing/src/node/nats.ts
// Short-term: avoid importing types from the `nats` package to prevent
// resolution errors during triage. Use permissive `any` aliases here and
// delegate runtime connection creation to the Overmind nats adapter.
type ConnectionOptions = any;
type NatsConnection = any;
type StreamConfig = any;
import {
	GenericContainer,
	type StartedTestContainer,
	Wait,
} from "testcontainers";

export interface NatsOptions {
	/** Docker image to use (default: nats:2.10-alpine) */
	image?: string;
	/** Enable JetStream (default: true) */
	enableJetStream?: boolean;
	/** Username for NATS authentication */
	username?: string;
	/** Password for NATS authentication */
	password?: string;
	/** Port to expose (0 = random) */
	port?: number;
}

export class NatsTestcontainer {
	private container: StartedTestContainer;
	private options: Required<
		Omit<NatsOptions, "username" | "password" | "port">
	> & {
		username?: string;
		password?: string;
		port?: number;
	};
	private _connection?: NatsConnection;

	private constructor(
		container: StartedTestContainer,
		options: Required<Omit<NatsOptions, "username" | "password" | "port">> & {
			username?: string;
			password?: string;
			port?: number;
		},
	) {
		this.container = container;
		this.options = options;
	}

	static async start(options: NatsOptions = {}): Promise<NatsTestcontainer> {
		const opts = {
			image: options.image ?? "nats:2.10-alpine",
			enableJetStream: options.enableJetStream ?? true,
			username: options.username,
			password: options.password,
			port: options.port,
		};

		const command: string[] = ["--name", "nats-testcontainer"];

		// Enable JetStream
		if (opts.enableJetStream) {
			command.push("-js");
		}

		// Add authentication if provided
		if (opts.username && opts.password) {
			command.push("--user", opts.username);
			command.push("--pass", opts.password);
		}

		const containerBuilder = new GenericContainer(opts.image)
			.withCommand(command)
			.withExposedPorts(opts.port ?? 4222)
			.withWaitStrategy(Wait.forLogMessage(/Server is ready/));

		const container = await containerBuilder.start();

		return new NatsTestcontainer(container, opts);
	}

	async stop(): Promise<void> {
		if (this._connection) {
			await this._connection.close();
			this._connection = undefined;
		}
		await this.container.stop();
	}

	async getClient(): Promise<NatsConnection> {
		if (!this._connection) {
			// Delegate to the overmind nats adapter which normalizes module shapes
			// across different installs and bundlers. Use a dynamic import to keep
			// the test helper lightweight when the adapter isn't needed.
			// Use require at runtime to avoid TypeScript trying to include files
			// from another package in the testing project's file list.
			// eslint-disable-next-line @typescript-eslint/no-var-requires
			const {
				connectNats,
			} = require("../../../overmind/src/clients/nats-adapter.js");

			const connectionOptions: ConnectionOptions = {
				servers: this.getConnectionString(),
			};

			if (this.options.username && this.options.password) {
				connectionOptions.user = this.options.username;
				connectionOptions.pass = this.options.password;
			}

			this._connection = await connectNats(connectionOptions);
		}
		return this._connection;
	}

	async createStream(config: StreamConfig): Promise<void> {
		const nc = await this.getClient();
		const jsm = await nc.jetstreamManager();
		await jsm.streams.add(config);
	}

	getConnectionString(): string {
		return `nats://${this.getHost()}:${this.getPort()}`;
	}

	getHost(): string {
		return this.container.getHost();
	}

	getPort(): number {
		return this.container.getMappedPort(this.options.port ?? 4222);
	}
}

File: packages/goblins/testing/src/python/nats.py
"""NATS JetStream Testcontainer for Python integration tests."""

from typing import Optional, Dict, Any
from testcontainers.core.container import DockerContainer
from testcontainers.core.waiting_utils import wait_for_logs
import nats
from nats.js.api import StreamConfig


class NatsContainer:
    """NATS JetStream container for integration testing with nats-py."""

    def __init__(
        self,
        image: str = "nats:2.10-alpine",
        enable_jetstream: bool = True,
        username: Optional[str] = None,
        password: Optional[str] = None,
        port: int = 4222
    ):
        self.image = image
        self.enable_jetstream = enable_jetstream
        self.username = username
        self.password = password
        self.port = port
        self._container: Optional[DockerContainer] = None
        self._connection: Optional[nats.NATS] = None

    async def start(self) -> "NatsContainer":
        """Start the NATS container."""
        command = ["--name", "nats-testcontainer"]

        if self.enable_jetstream:
            command.append("-js")

        if self.username and self.password:
            command.extend(["--user", self.username])
            command.extend(["--pass", self.password])

        self._container = DockerContainer(image=self.image)
        self._container.with_command(" ".join(command))
        self._container.with_exposed_ports(self.port)

        self._container.start()

        # Wait for NATS to be ready
        wait_for_logs(self._container, "Server is ready", timeout=30)

        return self

    async def stop(self) -> None:
        """Stop the NATS container."""
        if self._connection:
            await self._connection.close()
            self._connection = None

        if self._container:
            self._container.stop()
            self._container = None

    async def get_client(self) -> nats.NATS:
        """Get a NATS client connection."""
        if not self._connection:
            if not self._container:
                raise RuntimeError("Container not started. Call start() first.")

            self._connection = await nats.connect(
                servers=[self.get_connection_string()],
                user=self.username,
                password=self.password
            )

        return self._connection

    async def create_stream(self, config: StreamConfig) -> None:
        """Create a JetStream stream."""
        nc = await self.get_client()
        js = nc.jetstream()
        await js.add_stream(config)

    def get_connection_string(self) -> str:
        """Get the NATS connection string."""
        return f"nats://{self.get_host()}:{self.get_port()}"

    def get_host(self) -> str:
        """Get the container host."""
        if not self._container:
            raise RuntimeError("Container not started")
        return self._container.get_container_host_ip()

    def get_port(self) -> int:
        """Get the mapped port."""
        if not self._container:
            raise RuntimeError("Container not started")
        return int(self._container.get_exposed_port(self.port))

    async def __aenter__(self) -> "NatsContainer":
        """Async context manager entry."""
        return await self.start()

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Async context manager exit."""
        await self.stop()
